File: packages/sdk-ts/src/resources/secrets.ts
import type { EncliiClient } from '../client';
import type { Page, UnpagedListOptions } from '../types';
import type {
  BulkEnvVar,
  BulkSetEnvVarsResponse,
  EnvVar,
  SetEnvVarRequest,
} from '../types-ops';

/** Server-side cap on one bulk upsert (envvar_handlers.go BulkUpsertEnvVars). */
const MAX_BULK_VARIABLES = 100;

/**
 * Service environment variables and secrets
 * (`apps/switchyard-api/internal/api/envvar_handlers.go`).
 *
 * A variable is plaintext or a secret (`is_secret: true`). List and create
 * responses mask secret values as `••••••••`; `reveal()` returns the value and
 * the API records the reveal in its audit trail. Variables are addressed by
 * ID, not by key.
 */
export class SecretsResource {
  constructor(private readonly client: EncliiClient) {}

  /**
   * List a service's variables. `environment_id` narrows the list to one
   * environment; the endpoint returns every row in one response.
   */
  async list(
    serviceId: string,
    options: UnpagedListOptions & { environment_id?: string } = {},
  ): Promise<Page<EnvVar>> {
    const resp = await this.client.get<{
      environment_variables: EnvVar[] | null;
    }>(`/services/${encodeURIComponent(serviceId)}/env-vars`, {
      environment_id: options.environment_id,
    });
    return { data: resp.environment_variables ?? [], nextCursor: null };
  }

  /** Create a single variable. Set `is_secret: true` for secrets. */
  async set(serviceId: string, input: SetEnvVarRequest): Promise<EnvVar> {
    return this.client.post<EnvVar>(
      `/services/${encodeURIComponent(serviceId)}/env-vars`,
      input,
    );
  }

  /**
   * Create or update up to 100 variables by key in one call. `environment_id`
   * scopes the whole batch (omit for all environments). The API returns only
   * the number of variables written, not the variables.
   */
  async bulkSet(
    serviceId: string,
    vars: BulkEnvVar[],
    options: { environment_id?: string } = {},
  ): Promise<BulkSetEnvVarsResponse> {
    if (vars.length === 0 || vars.length > MAX_BULK_VARIABLES) {
      throw new Error(
        `secrets.bulkSet: expected 1 to ${MAX_BULK_VARIABLES} variables, got ${vars.length}`,
      );
    }
    return this.client.post<BulkSetEnvVarsResponse>(
      `/services/${encodeURIComponent(serviceId)}/env-vars/bulk`,
      {
        variables: vars,
        ...(options.environment_id
          ? { environment_id: options.environment_id }
          : {}),
      },
    );
  }

  /** Delete a variable by ID. */
  async delete(serviceId: string, varId: string): Promise<void> {
    await this.client.del(
      `/services/${encodeURIComponent(serviceId)}/env-vars/${encodeURIComponent(varId)}`,
    );
  }

  /**
   * Reveal the plaintext value of a secret. The call is logged for audit
   * and requires the developer role.
   */
  async reveal(
    serviceId: string,
    varId: string,
  ): Promise<{ key: string; value: string }> {
    return this.client.post<{ key: string; value: string }>(
      `/services/${encodeURIComponent(serviceId)}/env-vars/${encodeURIComponent(varId)}/reveal`,
      {},
    );
  }
}

File: packages/sdk-py/src/enclii_sdk/resources/secrets.py
"""Environment variable / secret management."""

from __future__ import annotations

from enclii_sdk.models.secrets import (
    CreateEnvVarRequest,
    EnvVarResponse,
    UpdateEnvVarRequest,
)
from enclii_sdk.resources._base import Resource


class SecretsResource(Resource):
    """CRUD for environment variables (both plaintext and secret)."""

    async def list(self, service_id: str) -> list[EnvVarResponse]:
        """List env vars for a service.

        API: ``GET /v1/services/{id}/env-vars``. Secret values are masked.
        """
        data = await self._client.get(f"/v1/services/{service_id}/env-vars")
        items = data if isinstance(data, list) else data.get("env_vars", [])
        return [EnvVarResponse.model_validate(v) for v in items]

    async def get(self, service_id: str, var_id: str) -> EnvVarResponse:
        """Fetch a single env var.

        API: ``GET /v1/services/{id}/env-vars/{var_id}``.
        """
        data = await self._client.get(f"/v1/services/{service_id}/env-vars/{var_id}")
        return EnvVarResponse.model_validate(data)

    async def create(
        self,
        service_id: str,
        *,
        key: str,
        value: str,
        is_secret: bool = False,
        environment_id: str | None = None,
    ) -> EnvVarResponse:
        """Create an env var. API: ``POST /v1/services/{id}/env-vars``."""
        request = CreateEnvVarRequest(
            key=key,
            value=value,
            is_secret=is_secret,
            environment_id=environment_id,
        )
        data = await self._client.post(
            f"/v1/services/{service_id}/env-vars",
            json_body=request.model_dump(mode="json", exclude_none=True),
        )
        return EnvVarResponse.model_validate(data)

    async def update(
        self,
        service_id: str,
        var_id: str,
        *,
        value: str | None = None,
        is_secret: bool | None = None,
    ) -> EnvVarResponse:
        """Update an env var. API: ``PUT /v1/services/{id}/env-vars/{var_id}``."""
        request = UpdateEnvVarRequest(value=value, is_secret=is_secret)
        data = await self._client.put(
            f"/v1/services/{service_id}/env-vars/{var_id}",
            json_body=request.model_dump(mode="json", exclude_none=True),
        )
        return EnvVarResponse.model_validate(data)

    async def delete(self, service_id: str, var_id: str) -> None:
        """Delete an env var. API: ``DELETE /v1/services/{id}/env-vars/{var_id}``."""
        await self._client.delete(f"/v1/services/{service_id}/env-vars/{var_id}")

    async def reveal(self, service_id: str, var_id: str) -> EnvVarResponse:
        """Reveal a secret's value (audit-logged).

        API: ``POST /v1/services/{id}/env-vars/{var_id}/reveal``. Requires
        ``developer``. Every call produces an audit entry.
        """
        data = await self._client.post(f"/v1/services/{service_id}/env-vars/{var_id}/reveal")
        return EnvVarResponse.model_validate(data)
